src/Core/IdentityServer/ResourceOwnerPasswordValidator.cs

```diff
@@ -135,5 +134,33 @@ protected async override Task<(User, bool)> ValidateContextAsync(ResourceOwnerPa
         {
             context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, customResponse: customResponse);
         }
+
+        private bool AuthEmailHeaderIsValid(ResourceOwnerPasswordValidationContext context)
+        {
+            if (!_currentContext.HttpContext.Request.Headers.ContainsKey("Auth-Email"))
+            {
+                return false;
+            }
+            else
+            {
+                try
+                {
+                    var authEmailHeader = _currentContext.HttpContext.Request.Headers["Auth-Email"];
+                    var authEmailDecoded = CoreHelpers.Base64UrlDecodeString(authEmailHeader);
+
+                    if (authEmailDecoded != context.UserName)
+                    {
+                        return false;
+                    }
+                }
+                catch (System.Exception e) when (e is System.InvalidOperationException || e is System.FormatException)
+                {
+                    // Invalid B64 encoding
+                    return false;
```

Comment:
Actually, after thinking through reverse compatibility for old clients, new server scenarios, we absolutely need to have a fallback check here to check if the original/old logic holds true:

```c#
return _currentContext.HttpContext.Request.Headers["Auth-Email"] == context.UserName;
```

Reply:
Never mind @eliykat , that's my bad, this method currently is not called since it's invocation is commented out, duh, it's early in the AM for me still, no coffee yet 😉